addons/controllers/antrea/antreaconfig_util.go

```diff
@@ -0,0 +1,95 @@
+package controllers
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	cniv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cni/v1alpha1"
+)
+
+// AntreaConfigSpec defines the desired state of AntreaConfig
+type antreaConfigSpec struct {
+	InfraProvider string `yaml:"infraProvider"`
+
+	Antrea antrea `yaml:"antrea,omitempty"`
+}
+
+type antrea struct {
+	AntConfig antConfig `yaml:"config,omitempty"`
+}
+
+type antConfig struct {
+	ServiceCIDR string `yaml:"serviceCIDR,omitempty"`
+
+	ServiceCIDRv6 string `yaml:"serviceCIDRv6,omitempty"`
+
+	TrafficEncapMode string `yaml:"trafficEncapMode,omitempty"`
+
+	NoSNAT bool `yaml:"noSNAT,omitempty"`
+
+	DisableUdpTunnelOffload bool `yaml:"disableUdpTunnelOffload,omitempty"`
+
+	DefaultMTU string `yaml:"defaultMTU,omitempty"`
+
+	TlsCipherSuites string `yaml:"tlsCipherSuites,omitempty"`
+
+	FeatureGates antreaFeatureGates `yaml:"featureGates,omitempty"`
+}
+
+type antreaFeatureGates struct {
+	AntreaProxy bool `yaml:"AntreaProxy,omitempty"`
+
+	EndpointSlice bool `yaml:"EndpointSlice,omitempty"`
+
+	AntreaPolicy bool `yaml:"AntreaPolicy,omitempty"`
+
+	FlowExporter bool `yaml:"FlowExporter,omitempty"`
+
+	Egress bool `yaml:"Egress,omitempty"`
+
+	NodePortLocal bool `yaml:"NodePortLocal,omitempty"`
+
+	AntreaTraceflow bool `yaml:"AntreaTraceflow,omitempty"`
+
+	NetworkPolicyStats bool `yaml:"NetworkPolicyStats,omitempty"`
+}
+
+func mapAntreaConfigSpec(cluster *clusterapiv1beta1.Cluster, config *cniv1alpha1.AntreaConfig) (*antreaConfigSpec, error) {
+	configSpec := &antreaConfigSpec{}
+
+	// Derive InfraProvider from the cluster
+	infraProvider, err := util.GetInfraProvider(cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to get InfraProvider")
+	}
+
+	configSpec.InfraProvider = infraProvider
+
+	// Derive ServiceCIDRs from the cluster
+	serviceCIDR, serviceCIDRv6, err := util.GetServiceCIDRs(cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to get serviceCIDR")
+	}
+	configSpec.Antrea.AntConfig.ServiceCIDR = serviceCIDR
+	configSpec.Antrea.AntConfig.ServiceCIDR = serviceCIDRv6
+	configSpec.Antrea.AntConfig.TrafficEncapMode = config.Spec.Antrea.AntConfig.TrafficEncapMode
+	configSpec.Antrea.AntConfig.NoSNAT = config.Spec.Antrea.AntConfig.NoSNAT
+	configSpec.Antrea.AntConfig.DisableUdpTunnelOffload = config.Spec.Antrea.AntConfig.DisableUdpTunnelOffload
+	configSpec.Antrea.AntConfig.DefaultMTU = config.Spec.Antrea.AntConfig.DefaultMTU
+	configSpec.Antrea.AntConfig.TlsCipherSuites = config.Spec.Antrea.AntConfig.TlsCipherSuites
+
+	// FeatureGates
+	configSpec.Antrea.AntConfig.FeatureGates.AntreaProxy = config.Spec.Antrea.AntConfig.FeatureGates.AntreaProxy
```

Comment:
According to https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/antrea/1.2.3/bundle/config/overlay/antrea_overlay.yaml#L101 when antreaProxy is enabled serviceCIDRs are not needed. Should we add CRD validation for this and not set serviceCIDR when this is true?

Reply:
I was thinking the same earlier after reading it, but the documentation also mentions that serviceCIDRs will be ignored when antreaProxy is enabled, so there should be no harm in setting them.
I added a Note to indicate that serviceCIDRs are automatically ignored.